### PR TITLE
Test implementation channels relevant to user csk 164

### PR DESCRIPTION
**Pull request by anton-pub**
This PR contains test implementation for:
https://pubnub.atlassian.net/browse/CSK-124  - Client Side Rate limiting
https://pubnub.atlassian.net/browse/CSK-166  - File send
https://pubnub.atlassian.net/browse/CSK-164  -

### CODE REVIEW

File: lib/tests/channel.test.ts [hunk 1]
```diff
@@ -723,5 +731,238 @@ describe("Channel test", () => {
...
+    await messageDraft.send()
+    await sleep(150) // Wait for the message to be sent and cached
+
+    const history = await channel.getHistory()
+
+    const messageInHistory = history.messages.find(
+      (message: any) => message.content.text === messageText
+    )
+
+    const extractedNamesFromText = extractMentionedUserIds(messageText)
```
Comment (piotr-suwala) [inline lib/tests/channel.test.ts:897]: I don't think you use this?

File: lib/tests/fixtures/example-video-oversize.mp4
Comment (piotr-suwala) [inline lib/tests/fixtures/example-video-oversize.mp4:1]: Do we actually use this file anywhere?

File: lib/tests/message.test.ts [hunk 2]
```diff
@@ -347,4 +355,610 @@ describe("Send message test", () => {
...
+    } catch (error) {
+      expect(error).toBeDefined()
+      expect(error.message).toContain("Message text cannot be empty")
+    }
+
+    try {
+      await channel.sendText("   ")
+    } catch (error) {
+      expect(error).toBeDefined()
+      expect(error.message).toContain("Message text cannot be empty")
```
Comment (piotr-suwala) [inline lib/tests/message.test.ts:737]: I don't think this test really enters this catch clock :P

Comment (piotr-suwala) [reply lib/tests/message.test.ts:737]: imho you should define a variable somewhere above and check the error message at the end of the test e.g.:
```
let errorMessage = ""
 try {
      await channel.sendText("   ")
    } catch (error) {
      errorMessage = error.message
    }
    
 expect(errorMessage).toContain("Message text cannot be empty")
```

File: lib/tests/message.test.ts [hunk 3]
```diff
@@ -347,4 +355,610 @@ describe("Send message test", () => {
...
+    await waitForAllMessagesToBeDelivered(messages, textMessages)
+
+    for (const textMessage of textMessages) {
+      expect(messages).toContain(textMessage)
+    }
+
+    disconnect()
+  }, 30000)
+  //Task created CSK-284
+  test("should fail to edit a deleted message", async () => {
```
Comment (piotr-suwala) [inline lib/tests/message.test.ts:778]: this test doesn't really check anything, right?

File: lib/tests/message.test.ts [hunk 4]
```diff
@@ -347,4 +355,610 @@ describe("Send message test", () => {
...
+    await sleep(150) // history calls have around 130ms of cache time
+
+    try {
+      await sentMessage.editText("Edited message")
+    } catch (error) {
+      expect(error.message).toContain("dfgdfgdfg not found")
+    }
+  }, 30000)
+
+  test("should delete toggle the message reaction", async () => {
```
Comment (piotr-suwala) [inline lib/tests/message.test.ts:796]: I don't understand this test

File: package.json [hunk 5]
```diff
@@ -30,13 +30,16 @@
     "@commitlint/config-conventional": "17.1.0",
     "@typescript-eslint/eslint-plugin": "5.37.0",
     "@typescript-eslint/parser": "5.37.0",
+    "axios": "^1.5.0",
+    "axios-mock-adapter": "^1.22.0",
```
Comment (piotr-suwala) [inline package.json:34]: whyyyy

File: samples/angular/src/chat-sdk/message-list/message-list.component.html [hunk 6]
```diff
@@ -0,0 +1,69 @@
+<div class="message-list">
+    <h1>Message List</h1>
+    <div>
+        <div>
+            <span>Channel pinned message: </span>
+            <span>{{ pinnedMessages[channel.id] }}</span>
+        </div>
+        <div *ngFor="let message of messages">
```
Comment (piotr-suwala) [inline samples/angular/src/chat-sdk/message-list/message-list.component.html:8]: rm this file

File: samples/angular/src/chat-sdk/message-list/message-list.component.ts [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+  MixedTextTypedElement,
+  ThreadChannel,
+  ThreadMessage,
+} from "@pubnub/chat"
+import { StateService } from "../../app/state.service"
+
+@Pipe({
+  name: "byPassSecurity",
+})
+export class ByPassSecurityPipe implements PipeTransform {
```
Comment (piotr-suwala) [inline samples/angular/src/chat-sdk/message-list/message-list.component.ts:16]: rm this file

File: samples/react-native-group-chat/components/message-text/MessageText.tsx [hunk 8]
```diff
@@ -33,6 +33,17 @@ export function MessageText({ onGoToMessage, messageProps }: MessageTextProps) {
     navigation.navigate("Chat")
   }
 
+  async function removeThread() {
+    if (messageProps.currentMessage?.originalPnMessage.hasThread) {
+      const thread = await messageProps.currentMessage?.originalPnMessage.getThread()
+      const messagesObject = await thread.getHistory()
```
Comment (piotr-suwala) [inline samples/react-native-group-chat/components/message-text/MessageText.tsx:39]: rm

File: lib/tests/channel.test.ts (hunk 1 above)
Comment (anton-pub) [reply lib/tests/channel.test.ts:897]: yep, thnx, cleaned up

File: lib/tests/message.test.ts (hunk 2 above)
Comment (anton-pub) [reply lib/tests/message.test.ts:737]: yep, thnx fixed

File: lib/tests/message.test.ts (hunk 3 above)
Comment (anton-pub) [reply lib/tests/message.test.ts:778]: yeah, I think I will come back to it after we will consider message delete behaviour. Meantime cleaned up.

File: lib/tests/message.test.ts (hunk 4 above)
Comment (anton-pub) [reply lib/tests/message.test.ts:796]: I fixed description. Mainly logic is the same like in should toggle the message reaction test just checking undo reaction

File: package.json (hunk 5 above)
Comment (anton-pub) [reply package.json:34]: thnx, cleaned up after test debugging

File: samples/angular/src/chat-sdk/message-list/message-list.component.html (hunk 6 above)
Comment (anton-pub) [reply samples/angular/src/chat-sdk/message-list/message-list.component.html:8]: ok removed

File: samples/angular/src/chat-sdk/message-list/message-list.component.ts (hunk 7 above)
Comment (anton-pub) [reply samples/angular/src/chat-sdk/message-list/message-list.component.ts:16]: ok removed

File: samples/react-native-group-chat/components/message-text/MessageText.tsx (hunk 8 above)
Comment (anton-pub) [reply samples/react-native-group-chat/components/message-text/MessageText.tsx:39]: ok removed

**Review by piotr-suwala — APPROVED** [review]
👍 🍏 👏

**Pull request merged**